middle_end/flambda2/simplify/simplify_set_of_closures.ml

```diff
@@ -578,24 +601,95 @@ let simplify_function0 context ~used_closure_vars ~shareable_constants
       ~dbg;
     decision
   in
+  let lifted_consts_this_function =
+    (* Subtle point: [uacc_after_upwards_traversal] must be used to retrieve all
+       of the lifted constants generated during the simplification of the
+       function, not [dacc_after_body]. The reason for this is that sometimes
+       the constants in [DA] are cleared (but remembered) and reinstated
+       afterwards, for example at a [Let_cont]. It follows that if the turning
+       point where the downwards traversal turns into an upwards traversal is in
+       such a context, not all of the constants may currently be present in
+       [DA]. *)
+    UA.lifted_constants uacc_after_upwards_traversal
+  in
+  let default_result_types =
+    Result_types.create_default ~params ~result_arity
+  in
+  let is_a_functor = Code.is_a_functor code in
+  let result_types =
+    if not (Flambda_features.function_result_types ~is_a_functor)
+    then default_result_types
+    else
+      match return_cont_uses with
+      | None -> default_result_types
```

Comment:
Doesn't it mean that the function never returns ? In this case we could actually have a `Bottom` result type, though maybe it's better to stay safe and use the default type anyway.

Reply:
Yes, I suppose that is the case...

Reply:
I've pushed a change for this, let's see how it goes.  I also added an `assert false` in a certain case.